### PR TITLE
Add Meter Style Selection Option

### LINKED ISSUE #1658: Allow meter style selection to be used in UI
**Has this feature been discussed and generally agreed?**
Not fully discussed and agreed yet, happy to receive comments.
Also it is a feature I would like to work on/implement.

**Describe the solution you'd like**
In short, this feature is to allow a user to select the meter style to be used indepently from the Skin selected.
Current there are 3 skins available which use the following settings:

_Normal_
In this skin the meter style defaults to a Bar

_Compact_
In this skin the meter style defaults to a Narrow Bar

_Fancy_
In this skin the meter style defaults to a set of LEDs

This new feature should allow the user to select the meter style to be used in the User Interface Settings options.
The selection will be made using a dropdown list.
It will no longer be connected to the selected skin.
Using the meter style should not increase the space used in the Compact skin when the Narrow Bar is used (the current mode for Compact skin).

Two new Meter Style options should be added:
One option to use Round LEDs (using the same size as the existing LEDs used for buffer/delay display)
Second option to use Small Round LEDs (using the same size as the Narrow Bar, mainly to save space in Compact mode)

**Implementation**
When implemented, this will be using a new setting in the ini file.
Once the feature is available, the default setting for this option will be based on the selected Skin.
When the selected skin is Fancy, the Meter Style is set to LEDs.
When the selected skin is Compact, the Meter Style is set to SLIM_BAR.
When the selected skin is Normal, the Meter Style is set to BAR.

### PR DESCRIPTION
**Pull request by henkdegroot**
This PR adds the option in the My Profile tab to allow the Meter Style type to be selected by the user.

Fixes #1658

### CODE REVIEW

**Comment by jujudusud** [comment]
Do we have to update screenshots for website ?

**Comment by jujudusud** [comment]
Do we have to update screenshots presents into the website ?

**Comment by softins** [comment]
> Do we have to update screenshots for website ?

I think that can wait until after the 3.8.0 release. The change in this PR will most likely not be included until after that, so will be in the following release.

**Comment by henkdegroot** [comment]
Converted to Draft to allow an extra meter style option to be added. Would it make sense to rebase the PR on master before pushing again?

**Comment by henkdegroot** [comment]
Rebased on current Master and resolved conflicts.

**Comment by pljones** [comment]
Keep an eye on the Clang-Format compliance checks.

**Comment by henkdegroot** [comment]
> Clang-Format compliance checks.

I actually did check them and at that moment I looked like the check failed in code which I have not changed....but checking again now and I see it also in code that I did change/add (I obvious did not run clang-format on my code before pushing again).

Will be updating this shortly and push again....

**Comment by ann0see** [comment]
Probably not in 3.8.1, I'm afraid.

**Comment by pljones** [comment]
Ooosch... wanted this to make it but I'd forgotten about it!

**Comment by pljones** [comment]
I'd like @gilgongo's views as there are UI/UX changes here.

Oh - I saw a strange bug earlier (not entirely related but I better write it down now before I totally forget).

image
In the Fancy skin, the meter is wide enough for MAG's little widget to show without wrapping.  Unfortunately........ in the Compact skin, it wraps.

This is with the Github build for:
```
38252048 2021-09-04 (HEAD -> master, upstream/master, origin/master, origin/HEAD) Unify JACK wording, fix bug, refactoring (#1889) [ann0see]
```
(so not _quite_ latest but it was when I ran it).

**Comment by henkdegroot** [comment]
> Oh - I saw a strange bug earlier (not entirely related but I better write it down now before I totally forget).

See #1451 , though I thought that was fixed/merged already....but it seems not.

**Comment by pljones** [comment]
> 
> 
> > Oh - I saw a strange bug earlier (not entirely related but I better write it down now before I totally forget).
> 
> See #1451 , though I thought that was fixed/merged already....but it seems not.

Raised #1993 with the fix.  Needs more work.

**Review by gilgongo — APPROVED** [review]
There are some minor glitches when changing through the different combinations, but mostly just a pixel getting pushed out here and there. Also notice that if small LEDs are chosen they don't seem to get used for the input meter sometimes? Not a big deal though.

**Comment by henkdegroot** [comment]
> Also notice that if small LEDs are chosen they don't seem to get used for the input meter sometimes? Not a big deal though.

This is by design. Compact settings apply to the mixer panel. It is something which should be detailed in the documentation (to prevent users registering issues for this as well).

The pixel shift, is that when changing from Fancy to a "Normal" view? Vaguely remember this might be caused by the difference in the slider.

**Comment by gilgongo** [comment]
> Compact settings apply to the mixer panel. It is something which should be detailed in the documentation (to prevent users registering issues for this as well).

Ah I see.

> The pixel shift, is that when changing from Fancy to a "Normal" view? Vaguely remember this might be caused by the difference in the slider.

Yes (I'm on Ubuntu 20.10 Gnome BTW). I think it may be unavoidable, and it's only in the transition - doesn't affect the view itself.

**Comment by gilgongo** [comment]
@jujudusud 

> Do we have to update screenshots for website ?

Yes, since the UI in settings has changed.

**Comment by henkdegroot** [comment]
> I'm hoping the UI will be self-explanatory

I would it is the same as for Skin. The different skins are also not specificallly shown on the website.

**Comment by gilgongo** [comment]
> I would it is the same as for Skin. The different skins are also not specificallly shown on the website.

Yes, maintaining text in different languages is hard enough without also doing multi-language screenshots so best to have screenshots only when necessary really.

**Comment by DavidSavinkoff** [comment]
Is there an option for 'no meter' ?

**Comment by henkdegroot** [comment]
> Is there an option for 'no meter' ?

Not at the moment. Is there a need/use case for that? It won't make the mixerboard "smaller" as the knob and client details still take the same amount of space.

**Comment by DavidSavinkoff** [comment]
Some things that I thought about having a no meter option:
- simple looking option, or need to rest eyes from flashing lights.
- low cpu load for low powered (or older) devices.
- mention the option before others do.

**Comment by gilgongo** [comment]
@DavidSavinkoff Certainly something to consider as part of the ongoing thinking about a new UI overall I think. In one of my explorations for a compact view I only had one LED flashing when the channel was making a sound (and indicating clipping when it happened). 

But that's not for this PR :-)

**Comment by pljones** [comment]
Also not directly related: another accessibilty point.  We've made it quite _visible_ on the UI when the mixer channel clips -- but there's no other notification.  I'd like there to be a (configurable, obviously) additional notification, both as a pop-up and as an audible bleep, ideally configurable per mixer channel, so you can turn it on for yourself and leave it off for everyone else.

**Comment by gilgongo** [comment]
> both as a pop-up and as an audible bleep

Which reminds me of this FR: https://github.com/jamulussoftware/jamulus/issues/425

**Comment by gilgongo** [comment]
Anyway, on the subject of **this** PR... I think we can merge?

**Review by pljones — APPROVED** [review]

**Comment by pljones** [comment]
Yeah, I don't see anything I don't like :).

I'm going to get conflicts on at least one of my work-in-progress branches... but that's my problem.

**Comment by pgScorpio** [comment]
I love the new meter styles, especially for use in the compact skin, but to be honest I would prefer a bar indicator with 3 colors. Is there a possibility for that too?

**Comment by pljones** [comment]
You mean the one that's currently just green?  Yeah...

The added advantage of the "LEDs" is that you can say, "Avoid the top one completely, only let the second one light occasionally, the next two mean 'good level' towards 'loud' and below that starts getting quiet very quickly".  With only colour to go on, you're relying on someone being happy with "green", "yellow" and "red".  For red/green colour blind, it might as well just be the single colour.

But having the option for those who can use it would be nice.

**Comment by pgScorpio** [comment]
> You mean the one that's currently just green? Yeah...

Yes I mean those we always had in the compact skin.

> The added advantage of the "LEDs" is that you can say, "Avoid the top one completely, only let the second one light occasionally, the next two mean 'good level' towards 'loud' and below that starts getting quiet very quickly". With only colour to go on, you're relying on someone being happy with "green", "yellow" and "red". For red/green colour blind, it might as well just be the single colour.

If you, like me, regularly work with very large groups (50-100 participants), and so using compact skin and multiple fader rows, the leds become somewhat hard to follow (so many blinking dots)..... The bars are easier to follow, but then you have no clear indication of the correct level...

> But having the option for those who can use it would be nice.

So Yes, it would be very nice!

**Comment by pgScorpio** [comment]
Just found another bug on compact skin:

Changing from Round LEDS's to Small LEDS's doesn't change the fader width. (But changing to Narrow Bar does.)

**Comment by henkdegroot** [comment]
> (But changing to Narrow Bar does.)

Correct noticed and this is as per design. The narrow bar is using the minimal space required. All LED options use a little more space. To prevent impact on appearance and usage, the size of the narrow bar has been kept the same as in the current version when using the compact skin. It is not intended that someone will keep switching between the different styles, and should therefore not be an issue.

> not logical when sorted on name in the list.

The current order of the list is the 3 existing options first and added new modes/options at the bottom.
This could be improved/changed but not sure if that is really needed.

> naming seems to be inconsistent

The Bar (Wide) sound incorrect to me, true that it is wider compared to Narrow but it is not wider compared to the LED's.
The LED's (Square) also seems not correct to me as they are not really square. Not sure what the correct "shape" name would be, rectangle would come closer to the actual shape. Currently just went for "without" any shape reference to make it a more "default" choice which matches the current versions used in Fancy/Normal/Compact skin.

**Comment by pljones** [comment]
> LED's (Square)

Could be "Stripe"?

**Comment by henkdegroot** [comment]
> Leds for Input L/R and 3 color bars for faders would be even nicer ;=)

Guess an improvement could be implemented to allow independent selection of the Input Meter style vs Mixerboard.

**Comment by pgScorpio** [comment]
> Guess an improvement could be implemented to allow independent selection of the Input Meter style vs Mixerboard.

That would be very appreciated!

**Comment by pgScorpio** [comment]
> > (But changing to Narrow Bar does.)
> 
> Correct noticed and this is as per design. The narrow bar is using the minimal space required. All LED options use a little > more space. To prevent impact on appearance and usage, the size of the narrow bar has been kept the same as in the > > current version when using the compact skin. It is not intended that someone will keep switching between the different > styles, and should therefore not be an issue.

But when changing from "small LED's" to "Round LED's" the fader width is adjusted, but when changing back to "small LED's" it stays wide. Only switching to "Narrow Bar" gives the narrow fader again. 
And "Small LED's" seem to fit into the same fader width as "Narrow Bar", as changing from "Narrow Bar" to "Small LED's" does not change the total fader width.

But when changing the meter type the fader width should always depend on the selected type, and it should not depend on a previous selected meter type. That's the bug I meant. 


> 
> > not logical when sorted on name in the list.
> 
> The current order of the list is the 3 existing options first and added new modes/options at the bottom. This could be improved/changed but not sure if that is really needed.

OK. But I think it would be more logical to group Bars and LED's in a logical order with a consistent naming format.


> 
> > naming seems to be inconsistent
> 
> The Bar (Wide) sound incorrect to me, true that it is wider compared to Narrow but it is not wider compared to the LED's. The LED's (Square) also seems not correct to me as they are not really square. Not sure what the correct "shape" name would be, rectangle would come closer to the actual shape. Currently just went for "without" any shape reference to make it a more "default" choice which matches the current versions used in Fancy/Normal/Compact skin.

I think Wide/Narrow and Big/Small is not really incorrect since they are relative, and you can't compare Wide to Big or Square, since width, size and shape are different types of attributes.
Also "LEDS's Square" could also be named "LED Bar" as this type of led VU meters are commonly referred to as such.

I was more referring to LED's/Bar without prefix/suffix and those with prefix/suffix, but I understand that's related to your previous remark about "the 3 existing options". 

So I still suggest changing the names for the official release, but this could also be:

Bar (Narrow)
Bar (Wide)
LED Bar        (or LED's (Bar) or Bar (LED's)  to be consistent in using a suffix)
LED's (Small)
LED's (Big)

or (Though I personally like the previous one better)

Narrow Bar
Wide Bar
LED Bar
Small LED's
Big LED's

**Comment by hoffie** [comment]
> But when changing from "small LED's" to "Round LED's" the fader width is adjusted, but when changing back to "small LED's" it stays wide. Only switching to "Narrow Bar" gives the narrow fader again.
> And "Small LED's" seem to fit into the same fader width as "Narrow Bar", as changing from "Narrow Bar" to "Small LED's" does not change the total fader width.
> But when changing the meter type the fader width should always depend on the selected type, and it should not depend on a previous selected meter type. That's the bug I meant.

I can confirm and have opened #2351 to track this.

**Comment by henkdegroot** [comment]
> But when changing from "small LED's" to "Round LED's" the fader width is adjusted, but when changing back to "small LED's" it stays wide. Only switching to "Narrow Bar" gives the narrow fader again. And "Small LED's" seem to fit into the same fader width as "Narrow Bar", as changing from "Narrow Bar" to "Small LED's" does not change the total fader width.
> 
> But when changing the meter type the fader width should always depend on the selected type, and it should not depend on a previous selected meter type. That's the bug I meant.

Thank you for spotting this, and thanks to @hoffie for already getting a fix for this.

**Comment by henkdegroot** [comment]
> So I still suggest changing the names for the official release, but this could also be:
> 
> Bar (Narrow) Bar (Wide) LED Bar (or LED's (Bar) or Bar (LED's) to be consistent in using a suffix) LED's (Small) LED's (Big)
> 
> or (Though I personally like the previous one better)
> 
> Narrow Bar Wide Bar LED Bar Small LED's Big LED's

Created #2353 for this.

**Comment by ann0see** [comment]
> Oh - I saw a strange bug earlier (not entirely related but I better write it down now before I totally forget).

Interesting: I also saw this happening with Qt6 #2363

**Comment by ann0see** [comment]
No it's about the white bars in the screenshot.
